pyteal/ast/opup.py

```diff
@@ -0,0 +1,163 @@
+                    "target_app_id must be specified in Explicit OpUp mode"
+                )
+            require_type(target_app_id, TealType.uint64)
+            self.target_app_id = target_app_id
+        elif mode == OpUpMode.OnCall:
+            self.target_app_id_slot = ScratchSlot()
+        else:
+            raise TealInputError("Invalid OpUp mode provided")
+
+        self.mode = mode
+
+        # A budget buffer is necessary to deal with an edge case of ensure_budget():
+        #   if the current budget is equal to or only slightly higher than the
+        #   required budget then it's possible for ensure_budget() to return with a
+        #   current budget less than the required budget. The buffer prevents this
+        #   from being the case.
+        self.buffer = Int(50)
+
+    def _create_app(self) -> Expr:
+        self.target_app_id = self.target_app_id_slot.load()
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.approval_program: ON_CALL_APP,
+                    TxnField.clear_state_program: ON_CALL_APP,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+            self.target_app_id_slot.store(Gitxn[0].created_application_id()),
+        )
+
+    def _delete_app(self, app_id: Expr) -> Expr:
+        return Seq(
+            InnerTxnBuilder.Begin(),
+            InnerTxnBuilder.SetFields(
+                {
+                    TxnField.type_enum: TxnType.ApplicationCall,
+                    TxnField.on_completion: OnComplete.DeleteApplication,
+                    TxnField.application_id: app_id,
+                }
+            ),
+            InnerTxnBuilder.Submit(),
+        )
+
+    def _ensure_budget_expr(self, required_budget: Expr) -> Expr:
+        return While(Global.opcode_budget() < required_budget).Do(
+            Seq(
+                InnerTxnBuilder.Begin(),
+                InnerTxnBuilder.SetFields(
+                    {
+                        TxnField.type_enum: TxnType.ApplicationCall,
+                        TxnField.application_id: self.target_app_id,
+                    }
+                ),
+                InnerTxnBuilder.Submit(),
+            )
+        )
+
+    def ensure_budget(self, required_budget: Expr) -> Expr:
+        """Ensure that the budget will be at least the required_budget."""
+        require_type(required_budget, TealType.uint64)
+
+        buffered_budget = required_budget + self.buffer
+        if self.mode == OpUpMode.OnCall:
+            return If(Global.opcode_budget() < buffered_budget).Then(
+                Seq(
+                    self._create_app(),
+                    self._ensure_budget_expr(buffered_budget),
+                    self._delete_app(self.target_app_id),
+                )
+            )
+
+        return self._ensure_budget_expr(buffered_budget)
```

Comment:
@algoidurovic For my understanding:  With `OpUpMode.Explicit`, what operations require a buffered budget?

Here's what I interpreted - let me know if/where I'm missing intuition:
*  `OpUpMode.Explicit` does _not_ require a buffered budget because control returns to the application _immediately_ after increasing opcode budget.
* `OpUpMode.OnCall` _does_ require a buffered budget because a delete app operation follows the opcode budget increase.
  * Aside:  How is `self.buffer = Int(50)` determined?
    * If I'm following, it seems the buffer can be specialized for the delete operation.

Reply:
Summarizing a verbal with @algoidurovic:
* I missed the following in my original note:  Since the budget is checked in a while loop, there's _always_ a check after op-ing up to see if the loop completed.  Consequently, there's _always_ a budget needed. 
* Igor to investigate:  Discussion highlighted that it'd probably make sense to invert the comparison ordering.  Concretely, switch from `Global.opcode_budget() < buffered_budget` to `buffered_budget >= Global.opcode_budget()`.  The change ensures `opcode_budget()` is interpreted as late as possible and provides the most up-to-date value.
* Igor to consider:  We discussed that the 50 opcode buffer is intended to be a rough, best fit for many use cases.  Since it's likely too large in some use cases, we discussed optionally offering a configurable override (e.g. factory method) for the app developer to supply a specific value.

Reply:
@algoidurovic Thanks for bringing in some feedback via https://github.com/algorand/pyteal/pull/274/commits/7152af649c6ab0d5a1a6690fcaee95268240b3ea.  

The remaining open question I see is:

> Igor to consider: We discussed that the 50 opcode buffer is intended to be a rough, best fit for many use cases. Since it's likely too large in some use cases, we discussed optionally offering a configurable override (e.g. factory method) for the app developer to supply a specific value.

Reply:
I added the option for the user to pass a custom buffer value to the OpUp constructor and tightened the default buffer value after some experimentation. Let me know if this addresses your concerns.

Reply:
@algoidurovic Thanks for bringing in the feedback.  

Before resolving, I have an _optional_ thought.  If you prefer as is, feel welcome to tell me and resolve the thread.

* Idea:  Since `budget` is _only_ referenced by `ensure_budget`, I'd optionally consider moving `buffer` from constructor parameter to method parameter.
* Rationale:  
  * Explicitly demarcates where `buffer` applies.  When provided to the constructor, the user must be aware of the implementation to see usage (e.g. Does `maximize_budget` use `buffer`?). 
  * Additionally, a user may wish to qualify `buffer` on a per invocation basis.  With a constructor parameter, the user must construct separate objects to express the use case.

Reply:
Resolving - As of https://github.com/algorand/pyteal/pull/274/commits/07bd6d25338c3f6da89b38dd9bfcebaceb777e34, `buffer` is no longer needed as a user-configurable parameter.